File: 2023/UICPC DIV1/UICPC Round #06/solutions & data/gwensgift/submissions/accepted/gift-db.py
#!/bin/python3

[n,k] = [int(x) for x in input().split()]
k -= 1

factorial = [1 for _ in range(n+1)]
for i in range(1,n+1):
    factorial[i] = factorial[i-1] * i

ans = []
perm = [0]
used = [False for _ in range(n)]

for i in range(n-1):
    possible = []
    for j in range(1,n):
        if not used[j]:
            possible.append( [(j-perm[i]) % n , j] )
    possible.sort()

    for x in possible:
        if factorial[n-i-2] <= k:
            k -= factorial[n-i-2]
            continue
        ans.append(x[0])
        perm.append(x[1])
        used[x[1]] = True
        break

print(*ans, sep=' ')

File: 2023/UICPC DIV1/UICPC Round #06/solutions & data/gwensgift/submissions/accepted/gift-db.cc
#include <bits/stdc++.h>
using namespace std;

const long long oo = LLONG_MAX / 2;

int main(){
  long long n, k;
  cin >> n >> k;
  k--;
  
  vector<long long> factorial(n);
  factorial[0] = 1;
  for(int i=1;i<n;i++){
    if(oo / factorial[i-1] > i) factorial[i] = factorial[i-1] * i;
    else factorial[i] = oo;
  }

  vector<int> ans = {0}; // Prepend 0 to the answer to make life easier...
  vector<int> perm = {0};
  vector<bool> used(n);
  for(int i=0;i<n-1;i++){
    vector<pair<int,int> > possible;
    for(int j=1;j<n;j++)
      if(!used[j])
	possible.emplace_back( (j-perm[i]+n) % n , j);
    
    sort(begin(possible), end(possible));
    
    for(auto x : possible){
      if(factorial[n-i-2] <= k) {
	k -= factorial[n-i-2];
	continue;
      }
      ans.push_back(x.first);
      perm.push_back(x.second);
      used[x.second] = true;
      break;
    }
  }
  
  for(int i=1;i<n;i++)
    cout << (i > 1 ? " " : "") << ans[i];
  cout << endl;

  // Sanity check...
  for(int i=1;i<n;i++)
    for(int j=i;j<n;j++)
      assert(accumulate(begin(ans)+i, begin(ans)+j+1, 0) % n);
}
